doc/_extensions/zephyr/domain.py

```diff
@@ -0,0 +1,287 @@
+    default_priority = 100
+
+    def apply(self):
+        matcher = NodeMatcher(CodeSampleNode)
+        for node in self.document.traverse(matcher):
+            self.convert_node(node)
+
+    def convert_node(self, node):
+        """
+        Transforms a `CodeSampleNode` into a `nodes.section` named after the code sample name.
+
+        Moves all sibling nodes that are after the `CodeSampleNode` in the documement under this new
+        section.
+        """
+        parent = node.parent
+        siblings_to_move = []
+        if parent is not None:
+            index = parent.index(node)
+            siblings_to_move = parent.children[index + 1:]
+
+            # TODO remove once all :ref:`sample-xyz` have migrated to :zephyr:code-sample:`xyz`
+            # as this is the recommended way to reference code samples going forward.
+            self.env.app.env.domaindata["std"]["labels"][node['id']] = self.env.docname, node['id'], node['name']
+            self.env.app.env.domaindata["std"]["anonlabels"][node['id']] = self.env.docname, node['id']
+
+            # Create a new section
+            new_section = nodes.section(ids=[node['id']])
+            new_section += nodes.title(text=node['name'])
+
+            # Move existing content from the custom node to the new section
+            new_section.extend(node.children)
+
+            # Move the sibling nodes under the new section
+            new_section.extend(siblings_to_move)
+
+            # Replace the custom node with the new section
+            node.replace_self(new_section)
+
+            # Remove the moved siblings from their original parent
+            for sibling in siblings_to_move:
+                parent.remove(sibling)
+
+class ProcessRelatedCodeSamplesNode(SphinxPostTransform):
+    default_priority = 5 # before ReferencesResolver
+
+    def run(self, **kwargs: Any) -> None:
+        matcher = NodeMatcher(RelatedCodeSamplesNode)
+        for node in self.document.traverse(matcher):
+            id = node['id'] # the ID of the node is the name of the doxygen group for which we
+                            # want to list related code samples
+
+            code_samples = self.env.domaindata["zephyr"]["code-samples"].values()
+            # Filter out code samples that don't reference this doxygen group
+            code_samples = [code_sample for code_sample in code_samples
+                            if id in code_sample['relevant-api']]
+
+            if len(code_samples) > 0:
+                admonition = nodes.admonition()
+                admonition += nodes.title(text="Related code samples")
+                admonition["classes"].append("related-code-samples")
+                sample_ul = nodes.bullet_list()
+                for code_sample in sorted(code_samples, key=lambda x: x['name']):
+                    # Create a list item that will eventually render as:
+                    # <li><a href="link-to-sample-id">sample-name</a> - sample-description</li>
```

Comment:
nit: Sphinx targets include others than HTML, so maybe this comment is not needed

Reply:
Ya when typing this I considered (and thought I had) formulating this "Create a list item that will eventually render as the following pseudo-html:". Happy to drop the comment altogether if it's not adding any value though.

Reply:
I dropped the comment. Hopefully the docutils API calls speak for themselves in the end :)